lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -0,0 +1,292 @@
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct EmittedFile {
+  VerbatimOp op;
+  StringAttr fileName;
+};
+
+struct BlackBoxReaderPass : public BlackBoxReaderBase<BlackBoxReaderPass> {
+  void runOnOperation() override;
+  void runOnAnnotation(Operation *op, DictionaryAttr anno, OpBuilder &builder);
+  void loadFile(Operation *op, StringRef inputPath, OpBuilder &builder);
+
+  using BlackBoxReaderBase::inputPrefix;
+  using BlackBoxReaderBase::resourcePrefix;
+
+private:
+  /// The generated `sv.verbatim` nodes.
+  SmallVector<EmittedFile, 8> emittedFiles;
+
+  /// The target directory to output black boxes into. Can be changed
+  /// through `firrtl.transforms.BlackBoxTargetDirAnno` annotations.
+  SmallString<128> targetDir;
+
+  /// The file list file name (sic) for black boxes. If set, generates a file
+  /// that lists all non-header source files for black boxes. Can be changed
+  /// through `firrtl.transforms.BlackBoxResourceFileNameAnno` annotations.
+  SmallString<128> resourceFileName;
+
+  // Internalized annotation class names for fast comparison.
+  StringAttr targetDirAnnoClass;
+  StringAttr resourceFileNameAnnoClass;
+  StringAttr inlineAnnoClass;
+  StringAttr pathAnnoClass;
+  StringAttr resourceAnnoClass;
+};
+} // end anonymous namespace
+
+/// Emit the annotated source code for black boxes in a circuit.
+void BlackBoxReaderPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  // Internalize some string attributes for easy reference later.
+  auto cx = &getContext();
+  targetDirAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxTargetDirAnno");
+  resourceFileNameAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceFileNameAnno");
+  inlineAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+  pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+  resourceAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
+
+  // Determine the target directory and resource file name from the
+  // annotations present on the circuit operation.
+  targetDir = ".";
+  resourceFileName = "firrtl_black_box_resource_files.f";
+
+  if (auto attrs = circuitOp->getAttrOfType<ArrayAttr>("annotations")) {
+    for (auto attr : attrs) {
+      auto anno = attr.dyn_cast<DictionaryAttr>();
+      if (!anno)
+        continue;
+      auto cls = anno.getAs<StringAttr>("class");
+
+      // Handle target dir annotation.
+      if (cls == targetDirAnnoClass) {
+        auto x = anno.getAs<StringAttr>("targetDir");
+        if (!x) {
+          circuitOp->emitError(targetDirAnnoClass.getValue() +
+                               " annotation missing \"targetDir\" attribute");
+          signalPassFailure();
+          continue;
+        }
```

Comment:
Thanks for being pedantic and checking this stuff and throwing good errors if mandatory fields are missing.

Reply:
:+1: I was thinking about maybe using the TableGen mechanism to generate `StructAttr` to do the parsing of the annotation dictionaries, but that just asserts if data is missing, rather than reporting back to the user. So I'm sticking with this at the moment, but it might be worthwhile contributing some code upstream that produces fallible casts for struct attrs that can properly report to the user if a dictionary should have a certain shape, but doesn't.

Reply:
Agree. This pattern shows up everywhere, e.g., [in Grand Central annotation parsing](https://github.com/llvm/circt/blob/main/lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp#L99). Asserts won't cut it as you want a real error message.

The other problem here is that the location really should be in the input JSON. However, there's no support that I'm aware of for getting source locators out of `llvm::json`. However, it would eventually be cool if we can include source locators pointing at JSON that can be used to generate top-tier error messages.

Reply:
Yeah that would be great to have!